ENH: signal.window: add array API support

jakevdp opened the pull request:
Towards gh-20678.

This PR adds [Array API standard](https://data-apis.org/array-api/) support to window function definitions within `scipy.signal`.

This is a bit different than other Array API changes, becuase these functions construct arrays from scratch. So rather than pulling the array API namespace from the inputs, this adds a new optional ``array_namespace`` keyword argument to each function that lets the user specify an array API; this defaults to ``scipy._lib.array_api_compat.numpy``, so there shouldn't be any user-visible change in the default case.

This should be considered a draft – I'm curious if maintainers have thoughts about whether this keyword argument is the best way to support the Array API for cases like this. Thanks!

lucascolley commented on scipy/signal/windows/_windows.py:
```diff
@@ -37,7 +45,7 @@ def _truncate(w, needed):
 
 
-def general_cosine(M, a, sym=True):
+def general_cosine(M, a, sym=True, *, array_namespace=None):
```
WDYT about naming the parameter `xp` instead? It is obviously a little more vague, but I feel that `array_namespace` is a bit too verbose for public API. That would align with what we already have in, e.g., [`fftfreq`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.fftfreq.html#scipy.fft.fftfreq)

lucascolley replied:
along the same comparison, perhaps we would want to add a `device` kwarg here too.

jakevdp replied:
Funny, I originally named the parameter `xp`, but thought it was a bit too vague for a public API, so I did a search-and-replace 😁

jakevdp replied:
Happy to change it back if you prefer that.

jakevdp replied:
Done

lucascolley commented on scipy/signal/windows/_windows.py:
```diff
@@ -14,6 +15,13 @@
 
 
+def _get_array_namespace(array_namespace):
+    if array_namespace is None:
+        return np_compat
```
In fft, we used the pattern `xp = np if xp is None else xp`. There is an argument to be made that the returned namespace should always be array API compatible. However, it is probably less disruptive to keep returning plain `np` for now. That probably makes more sense as a default, so that users aren't unexpectedly getting arrays from `array_api_compat` when they haven't enabled any array API stuff. Thoughts?

jakevdp replied:
I did that originally, but found that there are many array API functions that are not available in numpy 1.X (e.g. `np.cumulative_sum`, `np.acos`, np.acosh`, `np.concat`.

Originally I checked for these via `hasattr`, but then found the `np_compat` which seems to take care of these issues already.

I thought that `np_compat` was a cleaner solution than using `xp = np` and then overriding a bunch of things via `hasattr`, but I'm happy to switch back if you think the other approach would be cleaner.

jakevdp replied:
Maybe I'm misunderstanding what `np_compat` is: I thought it was `numpy` with some extra function definitions. Under what circumstance would these operations not result in a numpy array?

lucascolley replied:
No, I think I'm the one getting confused (it's easy with the complicated history involving `numpy.array_api`...). Since we are just returning arrays here, and the namespace used is an internal implementation detail, we are fine to use `np_compat` 👍. Resolving (maybe we would want to move this helper to `scipy/_lib/_array_api.py`).

lucascolley commented on scipy/signal/windows/_windows.py:
```diff
@@ -1548,40 +1605,40 @@ def chebwin(M, at, sym=True):
-        p = p * np.exp(1.j * np.pi / M * np.r_[0:M])
-        w = np.real(sp_fft.fft(p))
+        p = p * xp.exp(1.j * xp.pi / M * xp.arange(M))
+        w = xp.real(xp.fft.fft(p))
```
I think we should stick with `sp_fft` here, as that is array API compatible. As written, this is not fully portable, since `xp.fft` is only an optional extension.

jakevdp replied:
Done

lucascolley reviewed (commented):
the diff here looks fine to me, thanks! Would you like to start converting some of the tests? See gh-19001, or any of the recently merged `stats` PRs labelled with `array types` for examples.

We should let @rgommers weigh in on whether to name the parameter `array_namespace` or `xp`, and whether we would like a `device` parameter as well.

rgommers commented:
Thanks for this @jakevdp!

> whether to name the parameter `array_namespace` or `xp`, and whether we would like a `device` parameter as well.

We should do this consistently for all functions that create new arrays without receiving an array input. For [fftfreq](http://scipy.github.io/devdocs/reference/generated/scipy.fft.fftfreq.html) & co, we went with two keyword-only keywords: `xp=None, device=None`. Which still seems fine to me, so I suggest to go with that. `xp` is indeed very slightly non-descriptive, but I think if we use it consistently it will become quite recognizable to users soon enough.

rgommers commented:
I hadn't considered that argument for not considering the tests yet - you make a good point. So far I don't think we've accidentally introduced any regressions because of test changes. A separate PR seems like a potentially useful strategy, as long as it's available together with the code changes. Otherwise we have no way of testing the new functionality.

How about opening a second PR on top of this one including test changes? Then we can choose to either:
- merge this PR, then rebase and merge the second PR, or
- simply consider this one as thorough regression testing, and only merge the second PR when both are green (closing this PR)

lucascolley commented:
> How about opening a second PR on top of this one including test changes? Then we can choose to either:
> 
> merge this PR, then rebase and merge the second PR, or
> simply consider this one as thorough regression testing, and only merge the second PR when both are green (closing this PR)

That sounds like the right idea to me.

jakevdp committed:
ENH: scipy.signal.window: add array API support

jakevdp commented:
I updated `_windows.py` with the `xp` and `device` arguments, and squashed the commit. I'll plan to work on test changes in a second commit on top of that, probably some time next week.

jakevdp committed:
TST: array API support in test_windows

jakevdp commented:
I tried converting the test but it's failing, I think because the `xp` fixture tries to pass in `numpy`, which is not (currently) compatible with the Array API (e.g. `np.concat` doesn't exist). What's the best approach here?

lucascolley commented:
Right, this is why I didn't do this in `fftfreq`. It shouldn't be too complicated though, I think `xp_res = array_namespace(xp)` in the test functions, then passing `xp=xp_res` in to the functions being tested, should work.

jakevdp commented:
Should we worry about no longer having test coverage for the default call pattern, i.e. not passing `xp` at all?

lucascolley commented:
> Should we worry about no longer having test coverage for the default call pattern, i.e. not passing xp at all?

I would add a separate test for just that.

Kai-Striega commented on scipy/signal/windows/_windows.py:
```diff
@@ -53,6 +54,10 @@ def general_cosine(M, a, sym=True):
         When True (default), generates a symmetric window, for use in filter
         design.
         When False, generates a periodic window, for use in spectral analysis.
+    xp : module, optionaloptional array API namespace (default: numpy)
```
```suggestion
    xp : module, optional: array API namespace (default: numpy)
```

Just a minor typo

lucascolley commented:
(marked as draft until we have the PR with the test changes)

ev-br commented:
This has been merged as a part of gh-21783, closing as completed. 

Thank you @jakevdp for starting this, thanks to all reviewers.
